tensorrt_llm/serve/openai_disagg_service.py

```diff
@@ -0,0 +1,329 @@
+
+    async def setup(self) -> None:
+        self._ctx_client = self._client_factory(self._ctx_router, "ctx", self._config.max_retries)
+        self._gen_client = self._client_factory(self._gen_router, "gen", self._config.max_retries)
+
+        if self.disagg_cluster_config and self._cluster_storage:
+            logger.info("Starting disagg cluster manager")
+            self._disagg_cluster_manager = DisaggClusterManager(
+                self.disagg_cluster_config, self._cluster_storage
+            )
+            await self._disagg_cluster_manager.start()
+            await self._disagg_cluster_manager.watch_workers(on_event=self._on_worker_event)
+            logger.info("Disagg cluster manager started")
+        else:
+            if self._metadata_server and self._metadata_config:
+                logger.info("Starting server monitoring via metadata service")
+                await self._ctx_router.start_server_monitoring(
+                    self._metadata_config.refresh_interval
+                )
+                await self._gen_router.start_server_monitoring(
+                    self._metadata_config.refresh_interval
+                )
+            await self._wait_for_all_servers_ready()
+
+    async def teardown(self) -> None:
+        await self._ctx_client.shutdown()
+        await self._gen_client.shutdown()
+
+        if self._disagg_cluster_manager:
+            await self._disagg_cluster_manager.stop()
+
+        if self._metadata_server:
+            await self._ctx_router.stop_server_monitoring()
+            await self._gen_router.stop_server_monitoring()
+
+    async def _wait_for_all_servers_ready(self) -> None:
+        async def check_servers_ready():
+            elapsed_time = 0
+            interval = 3
+            while elapsed_time < self._server_start_timeout_secs:
+                _, unready_ctx_servers = await self._ctx_client.check_ready()
+                _, unready_gen_servers = await self._gen_client.check_ready()
+                if len(unready_ctx_servers) == 0 and len(unready_gen_servers) == 0:
+                    logger.info("All servers are ready")
+                    return
+                logger.info(
+                    f"Waiting for servers, context: {unready_ctx_servers}, generation: {unready_gen_servers}"
+                )
+                await asyncio.sleep(interval)
+                elapsed_time += interval
+
+        try:
+            await asyncio.wait_for(check_servers_ready(), timeout=self._server_start_timeout_secs)
+        except asyncio.TimeoutError:
+            raise TimeoutError("Timeout waiting for context and generation servers to be ready")
+
+    async def _on_worker_event(self, worker_info: WorkerInfo, event_type: WatchEventType):
+        router_map = {ServerRole.CONTEXT: self._ctx_router, ServerRole.GENERATION: self._gen_router}
+        worker_addr = f"{worker_info.host}:{worker_info.port}"
+        router = router_map[worker_info.role]
```

Comment:
In principle, `role` could also be `MM_ENCODER`. 
```suggestion
        router = router_map.get(worker_info.role)
        if router is None:
            logger.error(f"Unknown worker role: {worker_info.role}")
            return
```